### DISCUSSION TITLE
[@alpinejs/morph] - Usage on full HTML pages

### DISCUSSION THREAD

**Discussion by VicGUTT** [alpinejs/alpine#2511] [7. Bugs (archived)]
Starting this discussion following our [quick conversation on Twitter](https://twitter.com/VictorGutt/status/1471236768993366024?s=20) regarding using Alpine's morph plugin on full HTML pages and not simply on HTML partials/components.

[As previously stated](https://twitter.com/VictorGutt/status/1471275484419956740?s=20)

> [...] the morph plugin seems to be limited by it's use of `document.createRange().createContextualFragment` which, by creating a document fragment, seems incapable of returning the utter most parents (html & body)

Meaning, considering the following HTML structure:
```html
<html>
<head>
    <meta charset="UTF-8">
    <title>Hello</title>
</head>
<body>
    <p>Hey</p>
</body>
</html>
```

Alpine morph's [`createElement` implementation](https://github.com/alpinejs/alpine/blob/8d4f1266b25a550d9bd777b8aeb632a6857e89d1/packages/morph/src/morph.js#L64) will return:

```html
<meta charset="UTF-8">
```

and using `document.createRange().createContextualFragment` on the above HTML returns:

```html
▼ #document-fragment
    <meta charset="UTF-8">
    <title>Hello</title>
    <p>Hey</p>
```

which in turn causes the plugin to throw unexpected errors.

This behavior is to be opposed to [Morphdom's approache](https://github.com/patrick-steele-idem/morphdom/blob/84f28d0438f432cdef8cd869cd67ab916c029f2e/src/morphdom.js#L26,L36) of handling top level elements _(`#document`, `<html>` and `<body>`)_ differently than lower level elements.

Alpine's plugin behavior is similar to how Morphdom handles lower level elements.

Therefore, inspired by Morphdom, the plugin's `createElement` implementation could be refactored to something similar to:

```js
/**
 * @param {Element} from
 * @param {Element|DocumentFragment|string} to
 * @returns {Element}
 */
export default function createElement(from, to) {
    // return document.createRange().createContextualFragment(to).firstElementChild;

    const DOCUMENT_FRAGMENT_NODE = 11;

    if (['#document', 'HTML', 'BODY'].includes(from.nodeName) && typeof to === 'string') {
        const html = document.createElement('html');

        html.innerHTML = to;

        return html;
    }

    if (typeof to === 'string') {
        const template = document.createElement('template');

        template.innerHTML = to;

        return template.content.firstElementChild;
    }

    if (to.nodeType === DOCUMENT_FRAGMENT_NODE) {
        return to.firstElementChild;
    }

    return to;
}
```

I've created a playground/reproduction repository here: https://github.com/VicGUTT/alpine-morph-debug.
The commits of interest are:
- [fix: createElement refactor](https://github.com/VicGUTT/alpine-morph-debug/commit/66ee2a36644945662efe052441ac848a5a8100a0)
- [feat: simplified createElement](https://github.com/VicGUTT/alpine-morph-debug/commit/a97f10d6fec723dc96da8bcc9114c9e2f2d38268)

---

**Related discussion:** https://github.com/alpinejs/alpine/discussions/2514

**Comment by calebporzio** [comment]
Thanks for getting this conversation started!

If Alpine.morph were to be adapted to support this, what would your usage be? Like Turbolinks/drive functionality where you SPA-ify a server-driven app to change pages completely?

How would this handle the changing of <head> elements? (title, scripts, etc...)

Want to make sure what the end goal is and if this change doesn't fully satisfy that goal, maybe this functionality is better left to a different function/plugin alltogether?

**Comment by VicGUTT** [comment]
Yes, precisely for SPA-ifying a server-driven app. So handling full pages, but also partials/frames.
Basically a usage similar to:
```js
async function fetchAndMorphPage(morphOnlyFrames = []) {
    const htmlAsString = await fetch('some-page.html').then((res) => res.text());

    // No frames/partials ? Full page morph.
    if (!morphOnlyFrames.length) {
        await morph(document.documentElement, htmlAsString);

        return;
    }

    // Yes frames/partials ? Find, extract and morph only the target frames.

    const html = createContextualFragment(htmlAsString);

    for (const frame of morphOnlyFrames) {
        const from = document.querySelector(`#${frame}`);
        const to = html.querySelector(`#${frame}`);

        if (!from || !to) {
            continue;
        }

        await morph(from, to.outerHTML);
    }
}

function createContextualFragment(html) {
    return document.createRange().createContextualFragment(html);
}
```
**Some expectations:**
- First and foremost, at a high level, I do not expect Alpine.morph to behave any differently than Morphdom. My understanding is that it's meant to be Morphdom but Alpine aware. I may have gotten the wrong idea.
- I expect `<html>`, `<head>`, `<meta>`, `<title>`, `<body>` etc... to behave like any other DOM element _(`<div>`...)_.
- `<script>` & `<style>` should not behave any differently than when doing `document.documentElement.innerHTML = await fetch('/some-page').then(res => res.text());`. Meaning, the styles _(both inline and external)_ from `/some-page` should be applied and evaluated _(done automatically by browsers)_ while scripts _(both inline and external)_ should be ignored. Either way, I believe styles and scripts behaviors once morphed are out of scope for a **morph** package, as long as the package does not hinder the default browser behaviors _(which the plugin does not)_.

> Want to make sure what the end goal is and if this change doesn't fully satisfy that goal

This change seem to fully satisfy the goal of allowing top level elements.
However as stated in this discussion https://github.com/alpinejs/alpine/discussions/2514, there seem to be a different fault in the plugin's morphing algorithm _(I'd need to investigate more later)_ that results in unexpected behaviors. I don't believe that bug it tied to morphing full pages and could happen when usage is on simple components.

> maybe this functionality is better left to a different function/plugin alltogether?

It could be left to a different plugin, depending on what Alpine.morph's original goals and scopes are. But not enabling full pages seem like an arbitrary limitation of the plugin and is potentially the best use-case for the plugin in my opinion.

Ps.: Isn't one of the goals for the plugin to replace Livewire's forked Morphdom implementation ? Doesn't that warrant enabling this use-case ? _(I'm curious)_

**Reply by calebporzio** [reply]
Great points. Thanks.

Wanna PR the change?

**Reply by VicGUTT** [reply]
Definitely will. I should be able to submit the changes by Saturday.

One small question, after installing the project and running `npm run jest` there are no passing tests. All the expectations fail. I'm wondering if something might be wrong on my end _(I haven't looked much into it yet)_ or if indeed the y should be failing and fixed later ?

**Reply by pbowyer** [reply]
I had to search for the PR (as GitHub doesn't automatically create reverse-links to discussions - unlike with issues) so for fellow travellers, here's the PR https://github.com/alpinejs/alpine/pull/2529

**Reply by VicGUTT** [reply]
Ah yeah indeed, thanks for sharing the link @pbowyer.
If it's not too much of an ask, would you mind sharing your use-case for enabling this feature ?

**Reply by pbowyer** [reply]
@VicGUTT Currently, Morphdom compatibility - I found this issue when researching if morph was a drop-in replacement for morphdom. I don't have a use-case at present, but expect I'll want to diff full HTML pages (from legacy systems) sometime. Currently I'm receiving the whole HTML page and extracting sections and using Morph on them.

The `DOCUMENT_FRAGMENT_NODE` bit - when would I come across this? Not sure if I have already.

**Reply by VicGUTT** [reply]
Thanks for kindly sharing your _not present but possibly future_ use-case 😅.

It might be of interest to you to know I'm currently _(well, on and off since october actually)_ working on an Alpine plugin that aims to make the whole "SPA-like feeling" easier.
It's usage would be something like :

```html
<a href="/about" x-visit>About page</a>
<a href="/partials/some-partial" x-visit x-visit:only-frame="['partial-a', 'partial-b']">Some partial</a>
<button x-visit:post="'/contact'" x-visit:data="{ ... }" x-visit:headers="{ ... }">Send</button>
<form action="/contact" method="POST" x-visit>...</form>
<span x-visit="{ url: '/articles/123', method: 'PUT', data: { ... }, ... }">Update</span> <!-- Don't do this people 👀 -->
```

The [80/20 rule](https://en.wikipedia.org/wiki/Pareto_principle) _(the 80% of a certain piece of software can be written in 20% of the total allocated time. Conversely, the hardest 20% of the code takes 80% of the time)_ is definitely true on that one, where I technically just have some ironing out to do but it takes much of the time.

So if there is any interest from you or anyone passing by, it'd be ccol to discuss more on the subject :v:

---

> The `DOCUMENT_FRAGMENT_NODE` bit - when would I come across this? Not sure if I have already.

Ooh it is simply to check if the given element is a `#document-fragment` DOM node type. They are typically encountered when using `<template>` elements :

```js
const template = document.createElement('template');

console.log(template.content); // #document-fragment
console.log(template.content.nodeType); // 11
```

From the docs:
- https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template

Hope it helps.

**Comment by jimmywarting** [comment]
I just found out that morphing is going to be built in into the browsers soon if this lands across all browsers:  https://www.youtube.com/watch?v=JCJUPJ_zDQ4

**Reply by kpietraszko** [reply]
That's not DOM morphing, that's taking screenshots of before and after, and applying transformations (in the spatial sense) to transition. Seems totally irrelevant, sorry.

**Comment by doompadee** [comment]
Full page morph still does not work (with alpinejs/morph@3.13.2). As the provided PR has been [closed](https://github.com/alpinejs/alpine/issues/2529#issuecomment-1325712392_) without further details: is this use case out of scope of this project? Or would the current behavior count as a bug?

From my experiments I conclude that morphing individual elements within `<body>` seems to work. But morphing the `head` element does not work. It gets replaced with the first child of the new structure which is a show stopper.
